multicluster/controllers/multicluster/serviceexport_controller.go

```diff
@@ -311,40 +351,52 @@ func (r *ServiceExportReconciler) updateSvcExportStatus(ctx context.Context, req
 		return client.IgnoreNotFound(err)
 	}
 	now := metav1.Now()
-	var res, message *string
-	switch cause {
-	case notFound:
-		res = getStringPointer("not_found_service")
-		message = getStringPointer("the Service does not exist")
-	case importedService:
-		res = getStringPointer("imported_service")
-		message = getStringPointer("the Service is imported, not allowed to export")
-	default:
-		res = getStringPointer("invalid_service")
-		message = getStringPointer("the Service is not valid to export")
-	}
-
-	svcExportConditions := svcExport.Status.DeepCopy().Conditions
-	invalidCondition := k8smcsv1alpha1.ServiceExportCondition{
+	newCondition := k8smcsv1alpha1.ServiceExportCondition{
 		Type:               k8smcsv1alpha1.ServiceExportValid,
 		Status:             corev1.ConditionFalse,
 		LastTransitionTime: &now,
-		Reason:             res,
-		Message:            message,
 	}
 
-	matchedCondition := false
+	switch cause {
+	case serviceNotFound:
+		newCondition.Reason = getStringPointer("service_not_found")
+		newCondition.Message = getStringPointer("the Service does not exist")
+	case serviceWithoutEndpoints:
+		newCondition.Reason = getStringPointer("service_without_endpoints")
+		newCondition.Message = getStringPointer("the Service has no Endpoints, failed to export")
+	case isImportedService:
+		newCondition.Reason = getStringPointer("imported_service")
+		newCondition.Message = getStringPointer("the Service is imported, not allowed to export")
+	case serviceExported:
+		newCondition.Status = corev1.ConditionTrue
+		newCondition.Reason = getStringPointer("export_succeeded")
+		newCondition.Message = getStringPointer("the Service is exported successfully")
+	}
+
+	svcExportConditions := svcExport.Status.DeepCopy().Conditions
+	var existingCondition k8smcsv1alpha1.ServiceExportCondition
+	matchedConditionIdx := 0
 	for n, c := range svcExportConditions {
 		if c.Type == k8smcsv1alpha1.ServiceExportValid {
-			matchedCondition = true
-			svcExportConditions[n] = invalidCondition
+			existingCondition = c
+			matchedConditionIdx = n
 			break
 		}
 	}
-	if !matchedCondition {
-		svcExportConditions = append(svcExportConditions, invalidCondition)
 
+	if existingCondition != (k8smcsv1alpha1.ServiceExportCondition{}) {
+		if *existingCondition.Reason == *newCondition.Reason {
+			// No need to update the ServiceExport when there is no status change.
```

Comment:
As existingCondition is from cached client, this check may not be safe? But maybe not a big deal for status update.

Reply:
This reminds me why we need to get and check existing ResourceExport in updateOrCreateResourceExport()? It may get stale state from client cache too right? We should know it is update or not, from installedSvcs? Just in the restart case, we do not know as installedSvcs is empty, but we can check if ResourceExport exits or not, only for the Create case to cover that?

Reply:
The new updated ResourceExport will be set to use the same `ResourceVersion` from existing ResourceExport, if the cache is out of date, controller will fail to update the ResourceExport and retry.
We check the stored installed Service info to compare the information to determine if it needs to update or just skip updating ResourceExport. If controller restart, the update action will happen once without comparing due to emtpy installedSvcs.

Reply:
I saw `ResourceVersion`, but still I am asking is it necessary to get ResourceExport and compare, or in most cases we can just decide based on installedSvcs, but only when installedSvcs does not contain the Service we get ResourceExport?

If the Service is saved in installedSvcs, we know for sure an update is required, right? In that case, can we just update ResourceExport without get it first?

Reply:
We indeed compare installedSvcs instead of ResourceExport, the ResourceVersion is required when there is an update. It means we either store it in a cache or get it before update.

Reply:
I feel saving ResourceVersion to avoid possible update failure sounds better, but I am ok to keep the current approach too.